### PR TITLE
[website] Add a demo page for the Website

### PR DESCRIPTION
**Pull request by StrayBird-ATSH**
## Description ##
According to Issue (#688), a demo page is wanted in the GluonNLP website. So I am working on creating a demo web page for it. This is pull request will be continuously updated until the website is acceptable to be released.

### Essentials ###
- [ ] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

cc @dmlc/gluon-nlp-team

### CODE REVIEW

**Comment by mli** [comment]
Job PR-1180/4 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1180/4/index.html

**Comment by leezu** [comment]
```

code | 503
-- | --
type | "InternalServerException"
message | "Prediction failed"

```

File: docs/demo.rst [hunk 1]
```diff
@@ -0,0 +1,22 @@
+A Simple Demo for GluonNLP
```
Comment (eric-haibin-lin) [inline docs/demo.rst]: We may have more than 1 demo. We should have a "demo" category, and then with a drop down list with a few different demos:
- sentiment analysis
- question answering

File: docs/demo.rst [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+==========================
+
+Introduction
+------------
+
+This demo is using MMS Server
```
Comment (eric-haibin-lin) [inline docs/demo.rst]: Sentiment Analysis predicts whether an input is positive or negative. The model is based on BERT base, and are trained on the binary classification setting of the Stanford Sentiment Treebank. It achieves about 87% and 93.4% accuracy on the test set.

File: docs/demo.rst [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+Demo
+----
+
+Please input the following into the text box:
+
+.. raw:: html
+
+   ["Positive sentiment", "Negative sentiment"]
```
Comment (eric-haibin-lin) [inline docs/demo.rst]: We shall have a drop down list of a few pre-defined example inputs, and another text box to support custom user input.

**Comment by mli** [comment]
Job PR-1180/5 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1180/5/index.html

File: docs/demo/sentiment.rst [hunk 4]
```diff
@@ -0,0 +1,22 @@
+Sentiment Analysis
+==================
+
+Introduction
+------------
+
+Sentiment Analysis predicts whether an input is positive or negative. The model is based on BERT base, and are trained on the binary classification setting of the Stanford Sentiment Treebank. It achieves about 87% and 93.4% accuracy on the test set.
```
Comment (eric-haibin-lin) [inline docs/demo/sentiment.rst]: sorry, it should be "It achieves about 93.4% accuracy on the dev set. "

**Comment by mli** [comment]
Job PR-1180/9 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1180/9/index.html

**Comment by leezu** [comment]
> ```
> 
> code | 503
> -- | --
> type | "InternalServerException"
> message | "Prediction failed"
> ```

This may have been due to invalid input?

Currently the server just times out when submitting `["test"]` as a query.

**Comment by StrayBird-ATSH** [comment]
> > ```
> > 
> > code | 503
> > -- | --
> > type | "InternalServerException"
> > message | "Prediction failed"
> > ```
> 
> 
> This may have been due to invalid input?
> 
> 
> Currently the server just times out when submitting `["test"]` as a query.

I think it is due to invalid input. You may try out the default example I am now providing in the placeholder - it should work. Currently on the server side, I followed the same model as you provided in the #1140 .
I think the input arguments follow the same requirements you defined in that model. Maybe in the future we can provide more intuitive API or give more hints about which rules to follow when submitting inquiries.

**Comment by leezu** [comment]
> You may try out the default example I am now providing in the placeholder - it should work.

Still times out for me. How long do you wait for a response?

**Comment by StrayBird-ATSH** [comment]
> > You may try out the default example I am now providing in the placeholder - it should work.
> 
> Still times out for me. How long do you wait for a response?

Are you on Amazon Corporate VPN? VPN blocks 8888 port, so it will not work. I can get response within one second.

**Comment by leezu** [comment]
Yes, this was due to VPN.

**Comment by mli** [comment]
Job PR-1180/10 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1180/10/index.html

**Comment by StrayBird-ATSH** [comment]
Please help review whether the UI looks fine.
The JavaScript interaction of these controls are under developing and testing.

**Comment by mli** [comment]
Job PR-1180/12 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1180/12/index.html

**Comment by mli** [comment]
Job PR-1180/13 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1180/13/index.html

**Comment by StrayBird-ATSH** [comment]
[Help wanted] Currently I want to use an ajax to send the form information to the server and get the response so that I can display the information on the same page without refreshing the page. But I am stopped by the Cross Origin policy of ajax. It seems that modifying the server end is necessary for handling this problem? But the server running MMS is not a typical PHP/Java EE server, so I am not familiar about how to modify its response header.

 I am not sure I am seeking for the best strategy for solving this problem...

File: docs/demo/sentiment.rst [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+------------
+
+Sentiment Analysis predicts whether an input is positive or negative. The model is based on BERT base, and are trained on the binary classification setting of the Stanford Sentiment Treebank. It achieves about 93.4% accuracy on the dev set.
+
+Demo
+----
+
+Please input the following into the text box:
+
+   ["Positive sentiment", "Negative sentiment"]
```
Comment (eric-haibin-lin) [inline docs/demo/sentiment.rst]: We should just show one sentence here. Without quotes. Without brackets.

**Comment by mli** [comment]
Job PR-1180/14 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1180/14/index.html

**Comment by mli** [comment]
Job PR-1180/15 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1180/15/index.html

**Comment by StrayBird-ATSH** [comment]
This page is updated with AJAX feature, enabling displaying result without refreshing the page. Would you please have a look whether this page is OK or needs any enhancements?
Thanks! 
@eric-haibin-lin

File: docs/demo/sentiment.rst [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+Introduction
+------------
+
+Sentiment Analysis predicts whether an input is positive or negative. The model is based on BERT base, and are trained on the binary classification setting of the Stanford Sentiment Treebank. It achieves about 93.4% accuracy on the dev set.
+
+Demo
+----
+
+    You can either input a sentence into the textbox or select one sample from the select control.
```
Comment (eric-haibin-lin) [inline docs/demo/sentiment.rst]: Can you remove the space and special formatting for this instruction?

File: docs/demo/sentiment.rst [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+            -moz-appearance: none;
+            -webkit-appearance: none;
+            appearance: none;
+            background-image: url('data:image/svg+xml;charset=US-ASCII,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%22292.4%22%20height%3D%22292.4%22%3E%3Cpath%20fill%3D%22%23007CB2%22%20d%3D%22M287%2069.4a17.6%2017.6%200%200%200-13-5.4H18.4c-5%200-9.3%201.8-12.9%205.4A17.6%2017.6%200%200%200%200%2082.2c0%205%201.8%209.3%205.4%2012.9l128%20127.9c3.6%203.6%207.8%205.4%2012.8%205.4s9.2-1.8%2012.8-5.4L287%2095c3.5-3.5%205.4-7.8%205.4-12.8%200-5-1.9-9.2-5.5-12.8z%22%2F%3E%3C%2Fsvg%3E'),
+              linear-gradient(to bottom, #ffffff 0%,#ffffff 100%);
+            background-repeat: no-repeat, repeat;
+            background-position: right .7em top 50%, 0 0;
+            background-size: .65em auto, 100%;">
+      <option>Choose an example...</option>
+      <option value="Positive sentiment">Positive sentiment</option>
```
Comment (eric-haibin-lin) [inline docs/demo/sentiment.rst]: Can you replace this with some interesting/fun examples?

**Comment by mli** [comment]
Job PR-1180/16 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1180/16/index.html

File: docs/demo/sentiment.rst (hunk 7 above)
Comment (StrayBird-ATSH) [reply docs/demo/sentiment.rst]: This branch is updated. Would you please have a review and give some suggestions? Thanks!

**Comment by mli** [comment]
Job PR-1180/18 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1180/18/index.html

**Comment by StrayBird-ATSH** [comment]
@eric-haibin-lin This pull request should be ready now.

**Comment by mli** [comment]
Job PR-1180/19 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1180/19/index.html

**Comment by mli** [comment]
Job PR-1180/20 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1180/20/index.html

**Comment by StrayBird-ATSH** [comment]
> I tried the demo and did not see any response. Did it work for you?

Working now!

**Comment by StrayBird-ATSH** [comment]
> > I tried the demo and did not see any response. Did it work for you?
> 
> Working now!

@eric-haibin-lin @leezu

**Comment by mli** [comment]
Job PR-1180/21 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1180/21/index.html

**Comment by StrayBird-ATSH** [comment]
There has been constant failure of CI on cpu-unittest.
Would you please have a look at it? Thanks!
[Replay 20](https://ci.mxnet.io/blue/organizations/jenkins/GluonNLP-py3-cpu-unittest/detail/PR-1180/20/pipeline/)
[Replay 19](https://ci.mxnet.io/blue/organizations/jenkins/GluonNLP-py3-cpu-unittest/detail/PR-1180/19/pipeline/)
[Replay 18](https://ci.mxnet.io/blue/organizations/jenkins/GluonNLP-py3-cpu-unittest/detail/PR-1180/18/pipeline/)
[Replay 17](https://ci.mxnet.io/blue/organizations/jenkins/GluonNLP-py3-cpu-unittest/detail/PR-1180/17/pipeline/)
[Replay 16](https://ci.mxnet.io/blue/organizations/jenkins/GluonNLP-py3-cpu-unittest/detail/PR-1180/16/pipeline/)
[Replay 15](https://ci.mxnet.io/blue/organizations/jenkins/GluonNLP-py3-cpu-unittest/detail/PR-1180/15/pipeline/)
[Replay 14](https://ci.mxnet.io/blue/organizations/jenkins/GluonNLP-py3-cpu-unittest/detail/PR-1180/14/pipeline/)
@leezu

**Comment by mli** [comment]
Job PR-1180/22 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1180/22/index.html

**Comment by StrayBird-ATSH** [comment]
This pull request is ready for review and merge. Thanks!
@eric-haibin-lin

File: docs/demo/sentiment.rst [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+Introduction
+------------
+
+Sentiment Analysis predicts whether an input is positive or negative. The model is based on BERT base, and are trained on the binary classification setting of the Stanford Sentiment Treebank. It achieves about 93.4% accuracy on the dev set.
+
+Demo
+----
+
+You can either input a sentence into the textbox or select one sample from the select control.
```
Comment (leezu) [inline docs/demo/sentiment.rst:12]: Currently we're giving an existence proof of "it's possible to have some web application to do sentiment analysis". But to be really useful to the end users, we need to teach them how to build it (not only show it's possible).

To make things simple, such instructions should be based on https://aws.amazon.com/serverless/ and https://aws.amazon.com/sagemaker/

**Comment by sxjscience** [comment]
In fact, can we just provide a Colab/SageMaker Notebook?

**Comment by szha** [comment]
We should definitely provide colab/sagemaker links for all tutorials. Demo is different in that it show cases the model for certain task without any involvement of code. Tutorial for setting up serving/inference is indeed appealing to end users. We should definitely pursue that after this PR by summarizing the steps.

**Comment by sxjscience** [comment]
I think we can also provide demos with Colab + SageMaker, e.g., in https://magenta.tensorflow.org/demos/colab/. So we won't need to write any HTML codes (or just a few lines of HTML).

**Comment by leezu** [comment]
@sxjscience @szha I'm not sure just having a demo without providing any instructions on how to reproduce it (and how to deploy it) is very helpful without the demo meeting a very high bar. For a simple demo such as currently, we can make it special by showcasing how easy it is to deploy it on AWS.

**Comment by StrayBird-ATSH** [comment]
I am planning to add more descriptions on configuration and then add a CoLab. Do you think it is OK?

**Comment by szha** [comment]
@leezu users who just want to try our pre-trained models will benefit from the set up. These are people who don't concern themselves with understanding the code yet, so mixing the concern would create noise for this need.

**Comment by leezu** [comment]
@szha it doesn't need to be mixed. The proposal is to have an impressive demo, and provide information how to reproduce it showing the benefits of the technical stack. Not everyone has to read the latter part.

**Comment by szha** [comment]
Sounds good, @leezu. Do you have suggestions on how to improve the demo? Considering the investment, maybe the colab as @sxjscience suggested would indeed be the most efficient option.

Please continue the discussion and conclude the path forward.

**Comment by StrayBird-ATSH** [comment]
But actually our gpu-doc CI build is in faults. @leezu Maybe you can have a look at what is preventing our gpu-doc from being built? You may reference some latest builds like #1228 
So that I can make changes to the pages in this pull request and we can view the results.

**Comment by StrayBird-ATSH** [comment]
Since colab needs someone's Google account, should I use my personal Google account to create colab? And should we create some repo/directory specifically for colab? Or just give a static file in someone's Google Drive?

**Comment by sxjscience** [comment]
@StrayBird-ATSH I think we can do the same thing as d2l: automatically/manully uploading the notebooks to another repo called https://github.com/d2l-ai/d2l-en-colab. Then, when we are going to access specific notebooks, we can do this:
https://colab.research.google.com/github/d2l-ai/d2l-en-colab/blob/master/chapter_preliminaries/ndarray.ipynb. @szha @eric-haibin-lin What do you think?

**Comment by StrayBird-ATSH** [comment]
That sounds great! Do you have any ideas about where to place this repo? Or just give a folder in the current repo?

**Comment by StrayBird-ATSH** [comment]
> That sounds great! Do you have any ideas about where to place this repo? Or just give a folder in the current repo?

@eric-haibin-lin @szha

**Comment by sxjscience** [comment]
@szha Would you create a repo in DMLC?

**Comment by szha** [comment]
@sxjscience @StrayBird-ATSH I created https://github.com/dmlc/gluon-nlp-notebooks

How do we automate the deployment to that repo?

**Comment by sxjscience** [comment]
@szha I'm currently not sure about the details. Need to as the D2L people about the existing solution.

**Comment by szha** [comment]
Hi @astonzhang, how does the notebook deploy to colab work in d2l?

**Comment by StrayBird-ATSH** [comment]
> @sxjscience @StrayBird-ATSH I created https://github.com/dmlc/gluon-nlp-notebooks
> 
> How do we automate the deployment to that repo?

@szha 
Currently this repo is not initialized. Would you mind first initialize it (I think just adding a Readme will be enough) so that we can start contribute to this repo?

Thanks!

**Comment by szha** [comment]
@StrayBird-ATSH done

**Comment by StrayBird-ATSH** [comment]
Wonderful

**Comment by chenw23** [comment]
@leezu  Since I am currently using your trained model in this website and we are planning to create a colab. Maybe you can provide the python scripts that are equivalent to your trained models so that users can run the scripts and get the result in the colab?
